### PR TITLE
Enhancement: Add a MongoDB Motor Repository

### LINKED ISSUE #1874: Enhancement: Add a MongoDB Motor Repository or Plugin
### Discussed in https://github.com/orgs/litestar-org/discussions/1559

<div type='discussions-op-text'>

<sup>Originally posted by **Goldziher** April 20, 2023</sup>
I'd suggest we add a MongoDB repository to work directly with the Motor async driver. The issue with the driver is that it has no official types, just an unofficial (but very solid) stubs library. 

There are of course multiple ODMs such as beanie, Odmantic, Mongox that build on Motor + pydantic to create a smoother experience. I do think though that these have some inherent problems -

1. None of them uses the proper mongo query and aggregation syntax. They try to make it more `pythonic`, but i don't personally like that. 

2. They are not performant or performance orientated. 

3. They don't integrate as smoothly with us as a proper repository or plugin will. 

Thoughts and suggestions?</div>

<a href="https://polar.sh/litestar-org/litestar/issues/1874">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/1874/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Tsdevendra1**
### Pull Request Checklist

[//]: # "Please review the [Litestar contribution guidelines](https://github.com/litestar-org/litestar/blob/main/CONTRIBUTING.rst) for this repository."

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

### Description

This is a very very rough draft of implementing a MongoDB Motor Repository, I didn't want to go to far into it in-case I was going in the complete wrong direction. Would anyone mind confirming that this is generally correct?

### Close Issue(s)

Closes #1874

### CODE REVIEW

File: litestar/contrib/mongodb_motor/repository/_sync.py
Comment (Goldziher) [inline litestar/contrib/mongodb_motor/repository/_sync.py:1]: this can be autogenerated from the async code using the unasync pre-commit hook (ask @provinzkraut in discord on how to operate this).

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (Goldziher) [inline litestar/contrib/mongodb_motor/repository/_async.py:14]: pymongo can be used for sync code, but for async there is need for motor. You can use the types from there, no?

File: litestar/contrib/mongodb_motor/repository/_util.py
Comment (Goldziher) [inline litestar/contrib/mongodb_motor/repository/_util.py:26]: switch `exc` to `e` - its what we use elsewhere

File: litestar/contrib/mongodb_motor/repository/types.py
Comment (Goldziher) [inline litestar/contrib/mongodb_motor/repository/types.py:8]: so, why are you binding Document to a mutablemapping? you should actually bind this to a specific type that has all the methods associated with a document. AM i missing something?

File: litestar/contrib/mongodb_motor/repository/types.py
Comment (Goldziher) [inline litestar/contrib/mongodb_motor/repository/types.py:9]: This is redundant - you can simply import this type from the source where you need it, rather than obfuscate stuff by re-exporting

File: litestar/contrib/mongodb_motor/repository/types.py
Comment (Tsdevendra1) [reply litestar/contrib/mongodb_motor/repository/types.py:8]: I might be misunderstanding what you're saying here, apologies if I am, from what I understand the collection methods  such as `insert_one`, `insert_many`, take `Union[_DocumentType, RawBSONDocument]` where `_DocumentType = TypeVar("_DocumentType", bound=Mapping[str, Any])`. Should we support both `_DocumentType` and`RawBSONDocument` as inputs to the methods such as `add` and `add_many` on the repository?

Comment (Goldziher) [reply litestar/contrib/mongodb_motor/repository/types.py:8]: input maybe, but what about output? usually when you query a document you want to get a document instance, no?

Comment (Tsdevendra1) [reply litestar/contrib/mongodb_motor/repository/types.py:8]: I've not personally used pymongo + motor before so I'm not sure what the normal conventions are. It seems most of the `find_one`, `find` type methods return `_DocumentType`, is there some other type that is normally expected to be returned?

Comment (Goldziher) [reply litestar/contrib/mongodb_motor/repository/types.py:8]: no, this is it

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (JacobCoffee) [inline litestar/contrib/mongodb_motor/repository/_async.py:45]:
```suggestion
            The added dictionaries with the ``_id`` field set if the operation was successfully acknowledged.
```

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (JacobCoffee) [inline litestar/contrib/mongodb_motor/repository/_async.py:24]:
```suggestion
        """Add ``data`` to the collection.
```

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (JacobCoffee) [inline litestar/contrib/mongodb_motor/repository/_async.py:69]:
```suggestion
            **kwargs: Optional keyword arguments to pass to the underlying ``count_documents`` method.
```

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (JacobCoffee) [inline litestar/contrib/mongodb_motor/repository/_async.py:108]:
```suggestion
        """Build a query dictionary from the provided ``BeforeAfter`` filter values.
```

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (JacobCoffee) [inline litestar/contrib/mongodb_motor/repository/_async.py:130]:
```suggestion
            values: Values for ``IN`` clause.
```

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (JacobCoffee) [inline litestar/contrib/mongodb_motor/repository/_async.py:138]:
```suggestion
        """Build a query dictionary from the provided ``SearchFilter`` values.
```

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (JacobCoffee) [inline litestar/contrib/mongodb_motor/repository/_async.py:142]:
```suggestion
            value: Values for ``LIKE`` clause.
```

**Review by JacobCoffee — CHANGES_REQUESTED** [review]
single backticks -> double backticks for sphinx-doc to generate correctly. We will also need API reference documentation and usage docs

**Comment by provinzkraut** [comment]
Does it make sense to have this in the core package? I feel like this might be better suited for a separate package. @litestar-org/maintainers?

**Comment by Goldziher** [comment]
> Does it make sense to have this in the core package? I feel like this might be better suited for a separate package. @litestar-org/maintainers?

I think a mongoDB repository is solid in the core package - as long as its a repository. Reason being that mongo is the second most popular DB option after SQL DBs.

**Comment by Goldziher** [comment]
@Tsdevendra1 ?

**Comment by Tsdevendra1** [comment]
Sorry been a bit busy recently, will try get this PR finished and ready for review this week

**Comment by nathanielarking** [comment]
Hi @Tsdevendra1, thanks for working on this PR, I will defenitely find this useful. Let me know if you'd like any help - I haven't contributed before but I have a lot of time over the next few weeks to get into it.

**Comment by Tsdevendra1** [comment]
> Hi @Tsdevendra1, thanks for working on this PR, I will defenitely find this useful. Let me know if you'd like any help - I haven't contributed before but I have a lot of time over the next few weeks to get into it.

Thanks for the offer! Although I think all that's left now is generating the synchronous versions of the repository + tests, which can be automatically generated. I've asked one of the team for help with this, just waiting for a response and then will add it.

**Comment by Tsdevendra1** [comment]
@Goldziher just a quick update. I've implemented the sync version of the repo + tests. This PR now contains async + sync versions of the repo + tests for both. Is there anything else that needs to be added?

I'm not sure what the strategy is for actually ensuring that the repos work as intended. Manual testing? Or writing some integration tests?

**Comment by Goldziher** [comment]
> @Goldziher just a quick update. I've implemented the sync version of the repo + tests. This PR now contains async + sync versions of the repo + tests for both. Is there anything else that needs to be added?
> 
> I'm not sure what the strategy is for actually ensuring that the repos work as intended. Manual testing? Or writing some integration tests?

Tests

**Comment by Goldziher** [comment]
@gsakkis can you take point in reviewing and guiding this PR? You are our veritable mongo expert I think.

**Comment by gsakkis** [comment]
Heh far from a mongo expert but I'll take a look.

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (gsakkis) [inline litestar/contrib/mongodb_motor/repository/_async.py:391]: - All `_build_*query` methods return a dict with a single `field_name` key; they can be simplified by returning just the value.
- I'd just inline them, they're called only once and they're not that complex to be extracted into methods.

**Review by gsakkis — CHANGES_REQUESTED** [review]

File: litestar/contrib/mongodb_motor/repository/_util.py
Comment (provinzkraut) [reply litestar/contrib/mongodb_motor/repository/_util.py:35]: This serves a purpose actually. The sync version is auto-generated from the async one, but since the way to convert cursors to a list is different in sync vs async, a helper function is needed here. 

The sync version though is not needed, and can simply be replaced with `list`.

File: litestar/contrib/mongodb_motor/repository/_util.py
Comment (provinzkraut) [inline litestar/contrib/mongodb_motor/repository/_util.py:38]: This function can be removed. Just use `list` as a replacement directly.

File: litestar/contrib/mongodb_motor/repository/_util.py
Comment (provinzkraut) [reply litestar/contrib/mongodb_motor/repository/_util.py:35]: This serves a purpose actually. The sync version is auto-generated from the async one, but since the way to convertk cursors to a list is different in sync vs async, a helper function is needed here. 

The sync version though is not needed, and can simply be replaced with `list`.

File: litestar/contrib/mongodb_motor/repository/_async.py
Comment (Tsdevendra1) [reply litestar/contrib/mongodb_motor/repository/_async.py:373]: I'm not sure either, I couldn't figure out the equivalent based on the sqlalchemy repo so just left it as this for now

**Comment by provinzkraut** [comment]
@Tsdevendra1 Will be closing this one for now. We have decided it would be better suited as a standalone package. If you're interested in doing that, let us know and we'll help you port over your work there.

**Comment by Saphyel** [comment]
@Tsdevendra1 if you are doing the package please keep us update

**Pull request closed without merging**
